### PR TITLE
Add Apple Sign In strategy

### LINKED ISSUE #2: Add Apple OAuth 2 strategy
Based on https://github.com/danschultzer/pow_assent/issues/93

> iOS 13 has introduce [Apple OAuth2 auth](https://developer.apple.com/sign-in-with-apple/). It's just OAuth 2 with JWT so it'll be easy to implement. However, seems like Apple recommends using their JS for the request phase so maybe that is something that should be possible to configure with PowAssent?

### LINKED ISSUE #3: JWT logic
There is very weak JWT parsing in the current library. It would probably be best if there actually is some validation done on the JWT instead. It should either be a built-in module, or a third-party module that can be changed like with `:http_adapter`.

### PR DESCRIPTION
**Pull request by danschultzer**
Resolves #2

### CODE REVIEW

File: lib/assent/strategies/apple.ex [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+         {:ok, _private_key_id} <- Config.fetch(config, :private_key_id),
+         {:ok, private_key}     <- OAuth2.__load_private_key__(config) do
+
+      timestamp = DateTime.to_unix(DateTime.utc_now())
+      claims  = %{
+        "aud" => site,
+        "iss" => team_id,
+        "sub" => client_id,
+        "iat" => timestamp,
+        "exp" => timestamp + 600
```
Comment (yordis) [inline lib/assent/strategies/apple.ex]: Can we move this into a module attr and give it a name that represents the intention of the value?

File: lib/assent/strategies/apple.ex [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    config =
+      config
+      |> default_config()
+      |> Keyword.merge(config)
+
+    with {:ok, site}            <- Config.fetch(config, :site),
+         {:ok, client_id}       <- Config.fetch(config, :client_id),
+         {:ok, team_id}         <- Config.fetch(config, :team_id),
+         {:ok, _private_key_id} <- Config.fetch(config, :private_key_id),
+         {:ok, private_key}     <- OAuth2.__load_private_key__(config) do
```
Comment (yordis) [inline lib/assent/strategies/apple.ex]: What is the point of the private key id? I am confused that we don't do anything with it

File: lib/assent/strategies/oauth2.ex [hunk 3]
```diff
@@ -166,7 +166,7 @@ defmodule Assent.Strategy.OAuth2 do
   defp authentication_params(:private_key_jwt, config) do
     alg = Config.get(config, :jwt_algorithm, "RS256")
 
-    with {:ok, pem}             <- load_private_key(config),
+    with {:ok, pem}             <- __load_private_key__(config),
```
Comment (yordis) [inline lib/assent/strategies/oauth2.ex]: Could we remove these `__func_name__` functions?

Is there any reason why can't be named without `__`?

File: lib/assent/strategies/apple.ex (hunk 2 above)
Comment (danschultzer) [reply lib/assent/strategies/apple.ex]: Yeah, let me see if I can rewrite this. It's automatically included in the JWT signing process if it's available, so this is just to ensure that it has been defined as it's required for Apple Sign In. It's hard to read from this.

File: lib/assent/strategies/apple.ex (hunk 1 above)
Comment (danschultzer) [reply lib/assent/strategies/apple.ex]: I'm not sure if that's needed, I'm actually think about lowering the expiration. This is only meant for communicating with Apple Sign In, and it's intended to only be consumed once to fetch an access token.

File: lib/assent/strategies/oauth2.ex (hunk 3 above)
Comment (danschultzer) [reply lib/assent/strategies/oauth2.ex]: It's to keep it as a "private" API method. I didn't feel it was a good idea to expose it as public method for OAuth2. Probably makes more sense to move this to a helper module for JWT though.

File: lib/assent/strategies/apple.ex (hunk 1 above)
Comment (yordis) [reply lib/assent/strategies/apple.ex]: Right, could you give it a name that brings more clarity on the code? I don't know exactly why you adding 600 (or any other value) nor you can't document things based on a name from a module attr

For example:

```
@my_random_number_exp
```

If you mention that in your code, at least I could search for it.

So the idea is to make it more clear, for us.

**Pull request merged**